posix.mq_receive does not compile with cpp backend

### What happened?

This file compiles and runs with the C backend, but does not compile with the CPP backend.

```nim
# mqfail.nim
import posix

type Message = object
  value: int

const MQ_PATH = "/top_level_file"
const MQ_PRIORITY = 170
const MQ_MESSAGE_SIZE = sizeof(Message)

when isMainModule:
  var mqd_a: posix.MqAttr = MqAttr(mq_maxmsg: 10, mq_msgsize: clong(MQ_MESSAGE_SIZE))
  let writable: posix.Mqd = posix.mq_open(
    MQ_PATH,
    posix.O_CREAT or posix.O_WRONLY or posix.O_NONBLOCK,
    posix.S_IRWXU,
    addr(mqd_a)
  )
  let readable: posix.Mqd = posix.mq_open(
      MQ_PATH,
      posix.O_RDONLY or posix.O_NONBLOCK,
      posix.S_IRWXU,
      addr(mqd_a)
    )
  
  let sent: Message = Message(value: 88)
  block:
    let success: int = writable.mq_send(
      cast[cstring](sent.addr),
      MQ_MESSAGE_SIZE,
      MQ_PRIORITY
    )
    let error: cint = errno
    if success != 0:
      echo "Failed to write " &
        $sizeof(Message) &
        " bytes to mqd_t file descriptor " &
        $writable &
        " because failed with reason: " &
        $strerror(error)
    assert success == 0, $success
    
  block:
    var buffer: Message
    var priority: int
    let bytesRead: int = readable.mq_receive(cast[cstring](buffer.addr), MQ_MESSAGE_SIZE, priority)
    echo $buffer
    assert buffer == sent
    assert bytesRead == MQ_MESSAGE_SIZE
```

This code passes with compiled with `c` backend
```console
 nim c --verbosity:0 --run mqfail.nim
(value: 88)
```
but fails to compile with `cpp` backend:
```console
$ nim cpp --verbosity:0 --run mqfail.nim
/home/user/.cache/nim/mqfail_d/@mmqfail.nim.cpp: In function ‘void NimMainModule()’:
/home/user/.cache/nim/mqfail_d/@mmqfail.nim.cpp:319:104: error: cannot convert ‘NI*’ {aka ‘long int*’} to ‘unsigned int*’
  319 | il_18))), ((NI)8), (&priority__mqfail_19));
      |                    ~^~~~~~~~~~~~~~~~~~~~~
      |                     |
      |                     NI* {aka long int*}

In file included from /home/user/.cache/nim/mqfail_d/@mmqfail.nim.cpp:8:
/usr/include/mqueue.h:67:21: note:   initializing argument 4 of ‘ssize_t mq_receive(mqd_t, char*, size_t, unsigned int*)’
   67 |       unsigned int *__msg_prio) __nonnull ((2));
      |       ~~~~~~~~~~~~~~^~~~~~~~~~
Error: execution of an external compiler program 'g++ -c -std=gnu++17 -funsigned-char  -w -fmax-errors=3 -fpermissive -pthread   -I/home/user/n5m/lib -I/home/user -o /home/user/.cache/nim/mqfail_d/@mmqfail.nim.cpp.o /home/user/.cache/nim/mqfail_d/@mmqfail.nim.cpp' failed with exit code: 1
```

### Nim Version

f8b5464f31834bf97e7f9023644e255f2e9b6485
```console
$ nim -v
Nim Compiler Version 1.7.3 [Linux: amd64]
Compiled at 2022-10-31
Copyright (c) 2006-2022 by Andreas Rumpf

git hash: f8b5464f31834bf97e7f9023644e255f2e9b6485
active boot switches: -d:release
```

### Current Standard Output Logs

```shell
/home/user/.cache/nim/mqfail_d/@mmqfail.nim.cpp: In function ‘void NimMainModule()’:
/home/user/.cache/nim/mqfail_d/@mmqfail.nim.cpp:319:104: error: cannot convert ‘NI*’ {aka ‘long int*’} to ‘unsigned int*’
  319 | il_18))), ((NI)8), (&priority__mqfail_19));
      |                    ~^~~~~~~~~~~~~~~~~~~~~
      |                     |
      |                     NI* {aka long int*}

In file included from /home/user/.cache/nim/mqfail_d/@mmqfail.nim.cpp:8:
/usr/include/mqueue.h:67:21: note:   initializing argument 4 of ‘ssize_t mq_receive(mqd_t, char*, size_t, unsigned int*)’
   67 |       unsigned int *__msg_prio) __nonnull ((2));
      |       ~~~~~~~~~~~~~~^~~~~~~~~~
Error: execution of an external compiler program 'g++ -c -std=gnu++17 -funsigned-char  -w -fmax-errors=3 -fpermissive -pthread   -I/home/user/n5m/lib -I/home/user -o /home/user/.cache/nim/mqfail_d/@mmqfail.nim.cpp.o /home/user/.cache/nim/mqfail_d/@mmqfail.nim.cpp' failed with exit code: 1
```


### Expected Standard Output Logs

```shell
(value: 88)
```


### Possible Solution

Make Nim's `posix` module compatible with the types defined in [`mqueue.h`](https://man7.org/linux/man-pages/man3/mq_receive.3.html):
```c
ssize_t mq_receive(mqd_t mqdes, char *msg_ptr,
                   size_t msg_len, unsigned int *msg_prio);
```


### Additional Information

_No response_

--- a/lib/posix/posix.nim
+++ b/lib/posix/posix.nim
@@ -257,7 +257,7 @@ when not defined(nintendoswitch):
     importc, header: "<mqueue.h>".}
   proc mq_open*(a1: cstring, a2: cint): Mqd {.
     varargs, importc, header: "<mqueue.h>".} =
-    when defined(posix):
+    when defined(linux):
       runnableExamples:
         type Message = object
           value: int

--- a/lib/posix/posix.nim
+++ b/lib/posix/posix.nim
@@ -256,8 +256,55 @@ when not defined(nintendoswitch):
   proc mq_notify*(a1: Mqd, a2: ptr SigEvent): cint {.
     importc, header: "<mqueue.h>".}
   proc mq_open*(a1: cstring, a2: cint): Mqd {.
-    varargs, importc, header: "<mqueue.h>".}
-  proc mq_receive*(a1: Mqd, a2: cstring, a3: int, a4: var int): int {.
+    varargs, importc, header: "<mqueue.h>".} =
+    runnableExamples:
+      type Message = object
+        value: int
+
+      const MQ_PATH = "/top_level_file"
+      const MQ_PRIORITY = 170
+      const MQ_MESSAGE_SIZE = sizeof(Message)
+
+      var mqd_a: posix.MqAttr = MqAttr(mq_maxmsg: 10, mq_msgsize: clong(MQ_MESSAGE_SIZE))
+      let writable: posix.Mqd = posix.mq_open(
+        MQ_PATH,
+        posix.O_CREAT or posix.O_WRONLY or posix.O_NONBLOCK,
+        posix.S_IRWXU,
+        addr(mqd_a)
+      )
+      let readable: posix.Mqd = posix.mq_open(
+          MQ_PATH,
+          posix.O_RDONLY or posix.O_NONBLOCK,
+          posix.S_IRWXU,
+          addr(mqd_a)
+        )
+
+      let sent: Message = Message(value: 88)
+      block:
+        let success: int = writable.mq_send(
+          cast[cstring](sent.addr),
+          MQ_MESSAGE_SIZE,
+          MQ_PRIORITY
+        )
+        let error: cint = errno
+        if success != 0:
+          echo "Failed to write " &
+            $sizeof(Message) &
+            " bytes to mqd_t file descriptor " &
+            $writable &
+            " because failed with reason: " &
+            $strerror(error)
+        assert success == 0, $success
+
+      block:
+        var buffer: Message
+        var priority: cuint
+        let bytesRead: int = readable.mq_receive(cast[cstring](buffer.addr), MQ_MESSAGE_SIZE, priority)
+        echo $buffer
+        assert buffer == sent
+        assert bytesRead == MQ_MESSAGE_SIZE
+
+  proc mq_receive*(a1: Mqd, buffer: cstring, length: int, priority: var cuint): int {.
     importc, header: "<mqueue.h>".}
   proc mq_send*(a1: Mqd, a2: cstring, a3: int, a4: int): cint {.
     importc, header: "<mqueue.h>".}

--- a/lib/posix/posix.nim
+++ b/lib/posix/posix.nim
@@ -249,14 +249,14 @@ proc setlocale*(a1: cint, a2: cstring): cstring {.
 proc strfmon*(a1: cstring, a2: int, a3: cstring): int {.varargs,
    importc, header: "<monetary.h>".}
 
-when not defined(nintendoswitch):
+when not (defined(nintendoswitch) or defined(macos) or defined(macosx)):
   proc mq_notify*(mqdes: Mqd, event: ptr SigEvent): cint {.
     importc, header: "<mqueue.h>".}
 
   proc mq_open*(name: cstring, flags: cint): Mqd {.
     varargs, importc, header: "<mqueue.h>".} =
-    when defined(linux):
-      runnableExamples:
+    runnableExamples:
+      when defined(posix) and not (defined(nintendoswitch) or defined(macos) or defined(macosx)):
         type Message = object
           value: int
 

--- a/lib/posix/posix.nim
+++ b/lib/posix/posix.nim
@@ -257,52 +257,53 @@ when not defined(nintendoswitch):
     importc, header: "<mqueue.h>".}
   proc mq_open*(a1: cstring, a2: cint): Mqd {.
     varargs, importc, header: "<mqueue.h>".} =
-    runnableExamples:
-      type Message = object
-        value: int
+    when defined(posix):
+      runnableExamples:
+        type Message = object
+          value: int
 
-      const MQ_PATH = "/top_level_file"
-      const MQ_PRIORITY = 170
-      const MQ_MESSAGE_SIZE = sizeof(Message)
+        const MQ_PATH = "/top_level_file"
+        const MQ_PRIORITY = 170
+        const MQ_MESSAGE_SIZE = sizeof(Message)
 
-      var mqd_a: posix.MqAttr = MqAttr(mq_maxmsg: 10, mq_msgsize: clong(MQ_MESSAGE_SIZE))
-      let writable: posix.Mqd = posix.mq_open(
-        MQ_PATH,
-        posix.O_CREAT or posix.O_WRONLY or posix.O_NONBLOCK,
-        posix.S_IRWXU,
-        addr(mqd_a)
-      )
-      let readable: posix.Mqd = posix.mq_open(
+        var mqd_a: posix.MqAttr = MqAttr(mq_maxmsg: 10, mq_msgsize: clong(MQ_MESSAGE_SIZE))
+        let writable: posix.Mqd = posix.mq_open(
           MQ_PATH,
-          posix.O_RDONLY or posix.O_NONBLOCK,
+          posix.O_CREAT or posix.O_WRONLY or posix.O_NONBLOCK,
           posix.S_IRWXU,
           addr(mqd_a)
         )
+        let readable: posix.Mqd = posix.mq_open(
+            MQ_PATH,
+            posix.O_RDONLY or posix.O_NONBLOCK,
+            posix.S_IRWXU,
+            addr(mqd_a)
+          )
 
-      let sent: Message = Message(value: 88)
-      block:
-        let success: int = writable.mq_send(
-          cast[cstring](sent.addr),
-          MQ_MESSAGE_SIZE,
-          MQ_PRIORITY
-        )
-        let error: cint = errno
-        if success != 0:
-          echo "Failed to write " &
-            $sizeof(Message) &
-            " bytes to mqd_t file descriptor " &
-            $writable &
-            " because failed with reason: " &
-            $strerror(error)
-        assert success == 0, $success
+        let sent: Message = Message(value: 88)
+        block:
+          let success: int = writable.mq_send(
+            cast[cstring](sent.addr),
+            MQ_MESSAGE_SIZE,
+            MQ_PRIORITY
+          )
+          let error: cint = errno
+          if success != 0:
+            echo "Failed to write " &
+              $sizeof(Message) &
+              " bytes to mqd_t file descriptor " &
+              $writable &
+              " because failed with reason: " &
+              $strerror(error)
+          assert success == 0, $success
 
-      block:
-        var buffer: Message
-        var priority: cuint
-        let bytesRead: int = readable.mq_receive(cast[cstring](buffer.addr), MQ_MESSAGE_SIZE, priority)
-        echo $buffer
-        assert buffer == sent
-        assert bytesRead == MQ_MESSAGE_SIZE
+        block:
+          var buffer: Message
+          var priority: cuint
+          let bytesRead: int = readable.mq_receive(cast[cstring](buffer.addr), MQ_MESSAGE_SIZE, priority)
+          echo $buffer
+          assert buffer == sent
+          assert bytesRead == MQ_MESSAGE_SIZE
 
   proc mq_receive*(a1: Mqd, buffer: cstring, length: int, priority: var cuint): int {.
     importc, header: "<mqueue.h>".}

--- a/lib/posix/posix.nim
+++ b/lib/posix/posix.nim
@@ -250,38 +250,35 @@ proc strfmon*(a1: cstring, a2: int, a3: cstring): int {.varargs,
    importc, header: "<monetary.h>".}
 
 when not defined(nintendoswitch):
-  proc mq_close*(a1: Mqd): cint {.importc, header: "<mqueue.h>".}
-  proc mq_getattr*(a1: Mqd, a2: ptr MqAttr): cint {.
+  proc mq_notify*(mqdes: Mqd, event: ptr SigEvent): cint {.
     importc, header: "<mqueue.h>".}
-  proc mq_notify*(a1: Mqd, a2: ptr SigEvent): cint {.
-    importc, header: "<mqueue.h>".}
-  proc mq_open*(a1: cstring, a2: cint): Mqd {.
+
+  proc mq_open*(name: cstring, flags: cint): Mqd {.
     varargs, importc, header: "<mqueue.h>".} =
     when defined(linux):
       runnableExamples:
         type Message = object
           value: int
 
-        const MQ_PATH = "/top_level_file"
-        const MQ_PRIORITY = 170
-        const MQ_MESSAGE_SIZE = sizeof(Message)
+        const MQ_PATH: cstring = "/top_level_file"
+        const MQ_PRIORITY: cuint = 170
+        const MQ_MESSAGE_SIZE: csize_t = csize_t(sizeof(Message))
 
-        var mqd_a: posix.MqAttr = MqAttr(mq_maxmsg: 10, mq_msgsize: clong(MQ_MESSAGE_SIZE))
-        let writable: posix.Mqd = posix.mq_open(
-          MQ_PATH,
-          posix.O_CREAT or posix.O_WRONLY or posix.O_NONBLOCK,
-          posix.S_IRWXU,
-          addr(mqd_a)
+        let mqd_a: posix.MqAttr = MqAttr(
+          mq_maxmsg: 10,
+          mq_msgsize: clong(MQ_MESSAGE_SIZE)
         )
-        let readable: posix.Mqd = posix.mq_open(
+
+        block:
+          # send the message
+          let sent: Message = Message(value: 88)
+          let writable: posix.Mqd = posix.mq_open(
             MQ_PATH,
-            posix.O_RDONLY or posix.O_NONBLOCK,
+            posix.O_CREAT or posix.O_WRONLY or posix.O_NONBLOCK,
             posix.S_IRWXU,
             addr(mqd_a)
           )
 
-        let sent: Message = Message(value: 88)
-        block:
           let success: int = writable.mq_send(
             cast[cstring](sent.addr),
             MQ_MESSAGE_SIZE,
@@ -289,34 +286,65 @@ when not defined(nintendoswitch):
           )
           let error: cint = errno
           if success != 0:
-            echo "Failed to write " &
-              $sizeof(Message) &
-              " bytes to mqd_t file descriptor " &
-              $writable &
-              " because failed with reason: " &
-              $strerror(error)
-          assert success == 0, $success
+            echo $strerror(error)
 
         block:
+          # receive the message
+          let readable: posix.Mqd = posix.mq_open(
+            MQ_PATH,
+            posix.O_RDONLY or posix.O_NONBLOCK,
+            posix.S_IRWXU,
+            addr(mqd_a)
+          )
+
           var buffer: Message
           var priority: cuint
-          let bytesRead: int = readable.mq_receive(cast[cstring](buffer.addr), MQ_MESSAGE_SIZE, priority)
+          discard readable.mq_receive(
+            cast[cstring](buffer.addr),
+            MQ_MESSAGE_SIZE,
+            priority
+          )
           echo $buffer
-          assert buffer == sent
-          assert bytesRead == MQ_MESSAGE_SIZE
 
-  proc mq_receive*(a1: Mqd, buffer: cstring, length: int, priority: var cuint): int {.
-    importc, header: "<mqueue.h>".}
-  proc mq_send*(a1: Mqd, a2: cstring, a3: int, a4: int): cint {.
-    importc, header: "<mqueue.h>".}
-  proc mq_setattr*(a1: Mqd, a2, a3: ptr MqAttr): cint {.
+  proc mq_close*(mqdes: Mqd): cint {.importc, header: "<mqueue.h>".}
+
+  proc mq_receive*(
+    mqdes: Mqd,
+    buffer: cstring,
+    length: csize_t,
+    priority: var cuint
+  ): int {.importc, header: "<mqueue.h>".}
+
+  proc mq_timedreceive*(
+    mqdes: Mqd,
+    buffer: cstring,
+    length: csize_t,
+    priority: cuint,
+    timeout: ptr Timespec
+  ): int {.importc, header: "<mqueue.h>".}
+
+  proc mq_send*(
+    mqdes: Mqd,
+    buffer: cstring,
+    length: csize_t,
+    priority: cuint
+  ): cint {.importc, header: "<mqueue.h>".}
+
+  proc mq_timedsend*(
+    mqdes: Mqd,
+    buffer: cstring,
+    length: csize_t,
+    priority: cuint,
+    timeout: ptr Timespec
+  ): cint {.importc, header: "<mqueue.h>".}
+
+  proc mq_getattr*(mqdes: Mqd, attribute: ptr MqAttr): cint {.
     importc, header: "<mqueue.h>".}
 
-  proc mq_timedreceive*(a1: Mqd, a2: cstring, a3: int, a4: int,
-                        a5: ptr Timespec): int {.importc, header: "<mqueue.h>".}
-  proc mq_timedsend*(a1: Mqd, a2: cstring, a3: int, a4: int,
-                     a5: ptr Timespec): cint {.importc, header: "<mqueue.h>".}
-  proc mq_unlink*(a1: cstring): cint {.importc, header: "<mqueue.h>".}
+  proc mq_setattr*(mqdes: Mqd, newAttribute, oldAttribute: ptr MqAttr): cint {.
+    importc, header: "<mqueue.h>".}
+
+  proc mq_unlink*(mqdes: cstring): cint {.importc, header: "<mqueue.h>".}
 
 
 proc getpwnam*(a1: cstring): ptr Passwd {.importc, header: "<pwd.h>".}

--- a/lib/posix/posix.nim
+++ b/lib/posix/posix.nim
@@ -254,57 +254,7 @@ when not (defined(nintendoswitch) or defined(macos) or defined(macosx)):
     importc, header: "<mqueue.h>".}
 
   proc mq_open*(name: cstring, flags: cint): Mqd {.
-    varargs, importc, header: "<mqueue.h>".} =
-    runnableExamples:
-      when defined(posix) and not (defined(nintendoswitch) or defined(macos) or defined(macosx)):
-        type Message = object
-          value: int
-
-        const MQ_PATH: cstring = "/top_level_file"
-        const MQ_PRIORITY: cuint = 170
-        const MQ_MESSAGE_SIZE: csize_t = csize_t(sizeof(Message))
-
-        let mqd_a: posix.MqAttr = MqAttr(
-          mq_maxmsg: 10,
-          mq_msgsize: clong(MQ_MESSAGE_SIZE)
-        )
-
-        block:
-          # send the message
-          let sent: Message = Message(value: 88)
-          let writable: posix.Mqd = posix.mq_open(
-            MQ_PATH,
-            posix.O_CREAT or posix.O_WRONLY or posix.O_NONBLOCK,
-            posix.S_IRWXU,
-            addr(mqd_a)
-          )
-
-          let success: int = writable.mq_send(
-            cast[cstring](sent.addr),
-            MQ_MESSAGE_SIZE,
-            MQ_PRIORITY
-          )
-          let error: cint = errno
-          if success != 0:
-            echo $strerror(error)
-
-        block:
-          # receive the message
-          let readable: posix.Mqd = posix.mq_open(
-            MQ_PATH,
-            posix.O_RDONLY or posix.O_NONBLOCK,
-            posix.S_IRWXU,
-            addr(mqd_a)
-          )
-
-          var buffer: Message
-          var priority: cuint
-          discard readable.mq_receive(
-            cast[cstring](buffer.addr),
-            MQ_MESSAGE_SIZE,
-            priority
-          )
-          echo $buffer
+    varargs, importc, header: "<mqueue.h>".}
 
   proc mq_close*(mqdes: Mqd): cint {.importc, header: "<mqueue.h>".}
 

--- a/lib/posix/posix_macos_amd64.nim
+++ b/lib/posix/posix_macos_amd64.nim
@@ -108,15 +108,6 @@ type
     p_sign_posn*: char
     thousands_sep*: cstring
 
-  Mqd* {.importc: "mqd_t", header: "<mqueue.h>", final, pure.} = object
-  MqAttr* {.importc: "struct mq_attr",
-             header: "<mqueue.h>",
-             final, pure.} = object ## message queue attribute
-    mq_flags*: int   ## Message queue flags.
-    mq_maxmsg*: int  ## Maximum number of messages.
-    mq_msgsize*: int ## Maximum message size.
-    mq_curmsgs*: int ## Number of messages currently queued.
-
   Passwd* {.importc: "struct passwd", header: "<pwd.h>",
              final, pure.} = object ## struct passwd
     pw_name*: cstring   ## User's login name.

--- a/tests/stdlib/tposix.nim
+++ b/tests/stdlib/tposix.nim
@@ -19,7 +19,7 @@ when not defined(windows):
   writeLine(stdout, u.release)
   writeLine(stdout, u.machine)
 
-  block:
+  when not (defined(nintendoswitch) or defined(macos) or defined(macosx)):
     type Message = object
       value: int
 

--- a/tests/stdlib/tposix.nim
+++ b/tests/stdlib/tposix.nim
@@ -20,43 +20,44 @@ when not defined(windows):
   writeLine(stdout, u.machine)
 
   when not (defined(nintendoswitch) or defined(macos) or defined(macosx)):
-    type Message = object
-      value: int
-
-    const MQ_PATH: cstring = "/top_level_file"
-    const MQ_PRIORITY: cuint = 170
-    const MQ_MESSAGE_SIZE: csize_t = csize_t(sizeof(Message))
-
-    let mqd_a: posix.MqAttr = MqAttr(mq_maxmsg: 10, mq_msgsize: clong(MQ_MESSAGE_SIZE))
-    let writable: posix.Mqd = posix.mq_open(
-      MQ_PATH,
-      posix.O_CREAT or posix.O_WRONLY or posix.O_NONBLOCK,
-      posix.S_IRWXU,
-      addr(mqd_a)
-    )
-    let readable: posix.Mqd = posix.mq_open(
-      MQ_PATH,
-      posix.O_RDONLY or posix.O_NONBLOCK,
-      posix.S_IRWXU,
-      addr(mqd_a)
-    )
-
-    let sent: Message = Message(value: 88)
     block:
-      let success: int = writable.mq_send(
-        cast[cstring](sent.addr),
-        MQ_MESSAGE_SIZE,
-        MQ_PRIORITY
-      )
-      doAssert success == 0, $success
+      type Message = object
+        value: int
 
-    block:
-      var buffer: Message
-      var priority: cuint
-      let bytesRead: int = readable.mq_receive(
-        cast[cstring](buffer.addr),
-        MQ_MESSAGE_SIZE,
-        priority
+      const MQ_PATH: cstring = "/top_level_file"
+      const MQ_PRIORITY: cuint = 170
+      const MQ_MESSAGE_SIZE: csize_t = csize_t(sizeof(Message))
+
+      let mqd_a: posix.MqAttr = MqAttr(mq_maxmsg: 10, mq_msgsize: clong(MQ_MESSAGE_SIZE))
+      let writable: posix.Mqd = posix.mq_open(
+        MQ_PATH,
+        posix.O_CREAT or posix.O_WRONLY or posix.O_NONBLOCK,
+        posix.S_IRWXU,
+        addr(mqd_a)
       )
-      doAssert buffer == sent
-      doAssert bytesRead == int(MQ_MESSAGE_SIZE)
+      let readable: posix.Mqd = posix.mq_open(
+        MQ_PATH,
+        posix.O_RDONLY or posix.O_NONBLOCK,
+        posix.S_IRWXU,
+        addr(mqd_a)
+      )
+
+      let sent: Message = Message(value: 88)
+      block:
+        let success: int = writable.mq_send(
+          cast[cstring](sent.addr),
+          MQ_MESSAGE_SIZE,
+          MQ_PRIORITY
+        )
+        doAssert success == 0, $success
+
+      block:
+        var buffer: Message
+        var priority: cuint
+        let bytesRead: int = readable.mq_receive(
+          cast[cstring](buffer.addr),
+          MQ_MESSAGE_SIZE,
+          priority
+        )
+        doAssert buffer == sent
+        doAssert bytesRead == int(MQ_MESSAGE_SIZE)

--- a/tests/stdlib/tposix.nim
+++ b/tests/stdlib/tposix.nim
@@ -7,7 +7,7 @@ outputsub: ""
 when not defined(windows):
 
   import posix
-  import std/syncio
+  import std/[assertions, syncio]
 
   var
     u: Utsname
@@ -18,3 +18,45 @@ when not defined(windows):
   writeLine(stdout, u.nodename)
   writeLine(stdout, u.release)
   writeLine(stdout, u.machine)
+
+  block:
+    type Message = object
+      value: int
+
+    const MQ_PATH: cstring = "/top_level_file"
+    const MQ_PRIORITY: cuint = 170
+    const MQ_MESSAGE_SIZE: csize_t = csize_t(sizeof(Message))
+
+    let mqd_a: posix.MqAttr = MqAttr(mq_maxmsg: 10, mq_msgsize: clong(MQ_MESSAGE_SIZE))
+    let writable: posix.Mqd = posix.mq_open(
+      MQ_PATH,
+      posix.O_CREAT or posix.O_WRONLY or posix.O_NONBLOCK,
+      posix.S_IRWXU,
+      addr(mqd_a)
+    )
+    let readable: posix.Mqd = posix.mq_open(
+      MQ_PATH,
+      posix.O_RDONLY or posix.O_NONBLOCK,
+      posix.S_IRWXU,
+      addr(mqd_a)
+    )
+
+    let sent: Message = Message(value: 88)
+    block:
+      let success: int = writable.mq_send(
+        cast[cstring](sent.addr),
+        MQ_MESSAGE_SIZE,
+        MQ_PRIORITY
+      )
+      doAssert success == 0, $success
+
+    block:
+      var buffer: Message
+      var priority: cuint
+      let bytesRead: int = readable.mq_receive(
+        cast[cstring](buffer.addr),
+        MQ_MESSAGE_SIZE,
+        priority
+      )
+      doAssert buffer == sent
+      doAssert bytesRead == int(MQ_MESSAGE_SIZE)

